Source/WebCore/css/parser/CSSPropertyParser.cpp

```diff
@@ -6364,6 +6364,68 @@ bool CSSPropertyParser::consumeOffset(bool important)
     return m_range.atEnd();
 }
 
+bool CSSPropertyParser::consumeListStyleShorthand(bool important)
+{
+    auto& valuePool = CSSValuePool::singleton();
+    RefPtr<CSSValue> parsedPosition;
+    RefPtr<CSSValue> parsedImage;
+    RefPtr<CSSValue> parsedType;
+    unsigned noneCount = 0;
+
+    for (size_t i = 0; i < 3 && !m_range.atEnd(); ++i) {
+        if (m_range.peek().id() == CSSValueNone) {
+            ++noneCount;
+            consumeIdent(m_range);
+            continue;
+        }
+        if (!parsedPosition) {
+            if (auto position = parseSingleValue(CSSPropertyListStylePosition, CSSPropertyListStyle)) {
+                parsedPosition = position;
+                continue;
+            }
+        }
+        if (!parsedImage) {
+            if (auto image = parseSingleValue(CSSPropertyListStyleImage, CSSPropertyListStyle)) {
+                parsedImage = image;
+                continue;
+            }
+        }
+        if (!parsedType) {
+            if (auto type = parseSingleValue(CSSPropertyListStyleType, CSSPropertyListStyle)) {
+                parsedType = type;
+                continue;
+            }
+        }
+        return false;
+    }
+
+    if (!parsedImage && noneCount > 0) {
+        parsedImage = valuePool.createIdentifierValue(CSSValueNone);
+        --noneCount;
+    }
+
+    if (!parsedType && noneCount > 0) {
+        parsedType = valuePool.createIdentifierValue(CSSValueNone);
+        --noneCount;
+    }
+
+    if (noneCount > 0)
+        return false;
+
+    auto valueIsNone = [](RefPtr<CSSValue> value) -> bool {
+        return is<CSSPrimitiveValue>(value) && downcast<CSSPrimitiveValue>(*value).valueID() == CSSValueNone;
+    };
+    if (valueIsNone(parsedImage) && !parsedType) {
+        parsedImage = valuePool.createImplicitInitialValue();
```

Comment:
Why should `list-style: none` set `list-style-type: none` explicitly but `list-style-image: initial` implicitly?

Reply:
It's mostly a hack so `list-style: none` serializes to `list-style: none` instead of `list-style: none none`

Reply:
But it just seems strange to reset `parsedImage` just after setting it. Why not

```cpp
if (!parsedType && noneCount > 0) {
    parsedType = valuePool.createIdentifierValue(CSSValueNone);
    --noneCount;
}

if (!parsedImage && noneCount > 0) {
    parsedImage = valuePool.createIdentifierValue(CSSValueNone);
    --noneCount;
}

if (noneCount > 0)
    return false;

addPropertyWithImplicitDefault(CSSPropertyListStylePosition, CSSPropertyListStyle, WTFMove(parsedPosition), valuePool.createImplicitInitialValue(), important);
addPropertyWithImplicitDefault(CSSPropertyListStyleImage, CSSPropertyListStyle, WTFMove(parsedImage), valuePool.createImplicitInitialValue(), important);
addPropertyWithImplicitDefault(CSSPropertyListStyleType, CSSPropertyListStyle, WTFMove(parsedType), valuePool.createImplicitInitialValue(), important);
return m_range.atEnd();
```

Reply:
We do need to set list-style-type to none if list-style-type is not set, and list-style-image is none, this is what that code is doing.

Reply:
(FWIW setting longhand components to `initial` seems wrong in general, but that is a widespread problem in WebKit's shorthand parsing.)

Instead of setting `parsedImage` and then changing its value later, another approach could be to do something like this:

```c++
if (noneCount > (!!parsedImage + !!parsedType))
    return false;

if (noneCount == 2) {
    parsedImage = valuePool.createIdentifierValue(CSSValueNone);
    parsedType = valuePool.createIdentifierValue(CSSValueNone);
} else if (noneCount == 1) {
    // Use the implicit initial value for list-style-image if a single "none" was specified, to avoid serializing "none none".
    if (!parsedImage)
        parsedImage = parsedType ? valuePool.createIdentifierValue(CSSValueNone) : valuePool.createImplicitInitialValue();
    if (!parsedType)
        parsedType = valuePool.createIdentifierValue(CSSValueNone);
}
```